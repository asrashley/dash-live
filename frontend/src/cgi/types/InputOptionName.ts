export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}