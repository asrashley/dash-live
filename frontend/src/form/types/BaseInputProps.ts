import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

