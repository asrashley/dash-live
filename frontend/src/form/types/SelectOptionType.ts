export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
