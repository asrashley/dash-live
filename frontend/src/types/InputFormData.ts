export type InputFormData = {
    [name: string]: InputFormData | string | number | boolean | undefined;
};