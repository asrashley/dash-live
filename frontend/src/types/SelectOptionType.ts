export interface SelectOptionType {
    disabled?: boolean;
    name?: string;
    title: string;
    value: string;
}
