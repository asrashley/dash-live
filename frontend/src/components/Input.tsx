import { type ReadonlySignal, useComputed } from '@preact/signals';

import { DataListInput } from './DataListInput';
import { MultiSelectInput } from './MultiSelectInput';
import { RadioInput } from './RadioInput';
import { SelectInput } from './SelectInput';
import type { SetValueFunc } from '../types/SetValueFunc';
import type { InputFormData } from '../types/InputFormData';
import type { FormInputItem } from '../types/FormInputItem';
import type { FormRowMode } from '../types/FormRowMode';
import { FormGroupsProps } from '../types/FormGroupsProps';

export interface InputProps extends FormInputItem {
  data: FormGroupsProps['data'];
  disabledFields: FormGroupsProps['disabledFields'];
  mode: FormRowMode;
  validation?: "was-validated" | "has-validation" | "needs-validation";
  setValue: SetValueFunc;
}

type BaseInputProps = {
  name: string;
  id: string;
  type: FormInputItem["type"];
  className: string;
  title: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

interface GetNameProps {
  cgiName: string;
  mode: FormRowMode;
  shortName: string;
  fullName: string;
  prefix?: string;
}

function getName({ cgiName, mode, shortName, fullName, prefix }: GetNameProps): string {
  switch(mode) {
    case 'cgi':
      return cgiName;
    case 'shortName':
      return shortName;
    case 'fullName':
      if (prefix) {
        return `${prefix}__${fullName}`;
      }
      return fullName;
  }
}

export function Input({
  className = "",
  datalist_type,
  type,
  name: cgiName,
  shortName,
  fullName,
  prefix,
  mode,
  data,
  disabledFields,
  setValue,
  error,
  validation,
  title,
  options,
}: InputProps) {
  const name = getName({mode, shortName, cgiName, fullName, prefix});
  const value = useComputed<string | number | undefined>(() => {
      if (!data.value) {
        return undefined;
      }
      const val = (mode === 'fullName' && prefix) ? data.value[prefix][fullName] : data.value[name];
      if (typeof val === "boolean") {
        return val ? "1": "0";
      }
      return val;
  });
  const prefixData = useComputed<InputFormData>(() => (mode === 'fullName' && prefix) ? (data.value[prefix] as InputFormData): data.value);
  const stringValue = useComputed<string>(() => value.value === undefined ? "" : typeof value.value === "string" ? value.value : `${value.value}`);
  const disabled = useComputed<boolean>(() => !!disabledFields[name]);
  const inputClass =
    type === "checkbox"
      ? "form-check-input"
      : type === "select"
      ? "form-select"
      : "form-control";
  const validationClass = error
    ? " is-invalid"
    : validation === "was-validated"
    ? " is-valid"
    : "";
  const inpProps: BaseInputProps = {
    name,
    type,
    title,
    disabled,
    className: `${inputClass}${validationClass} ${className}`,
    id: `model-${name}`,
    "aria-describedby": `field-${name}`,
    onInput: (ev: Event) => {
      const target = ev.target as HTMLInputElement;
      setValue(name, target.type === "checkbox" ? target.checked : target.value);
    },
  };
  switch (type) {
    case "radio":
      return <RadioInput setValue={setValue} value={stringValue} options={options} disabledFields={disabledFields} {...inpProps} />;
    case "select":
      return <SelectInput setValue={setValue} value={stringValue} options={options} {...inpProps} />;
    case "multiselect":
      return <MultiSelectInput setValue={setValue} data={prefixData} options={options} {...inpProps} />;
    case "datalist":
      inpProps.type = datalist_type ?? 'text';
      return <DataListInput options={options} value={stringValue} {...inpProps} />;
    case 'checkbox':
      return <input checked={value.value === "1"} {...inpProps} />;
    default:
      return <input {...inpProps} />;
  }
}
