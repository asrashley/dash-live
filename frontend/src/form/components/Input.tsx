import { useComputed } from '@preact/signals';

import { DataListInput } from './DataListInput';
import { MultiSelectInput } from './MultiSelectInput';
import { PasswordInput } from './PasswordInput';
import { RadioInput } from './RadioInput';
import { SelectInput } from './SelectInput';
import type { InputFormData } from '../types/InputFormData';
import type { FormRowMode } from '../types/FormRowMode';
import { BaseInputProps } from '../types/BaseInputProps';
import { InputProps } from '../types/InputProps';


interface GetNameProps {
  cgiName: string;
  mode: FormRowMode;
  shortName?: string;
  fullName?: string;
  prefix?: string;
}

function getName({ cgiName, mode, shortName, fullName, prefix }: GetNameProps): string {
  switch(mode) {
    case 'cgi':
      return cgiName;
    case 'shortName':
      return shortName ?? cgiName;
    case 'fullName':
      if (prefix) {
        return `${prefix}__${fullName ?? cgiName}`;
      }
      return fullName ?? cgiName;
  }
}

export function Input({
  allowReveal,
  className = "",
  datalist_type,
  type,
  name: cgiName,
  shortName,
  fullName,
  prefix,
  mode = "cgi",
  describedBy,
  data,
  disabledFields,
  error,
  setValue,
  title,
  options,
  placeholder
}: InputProps) {
  const name = getName({mode, shortName, cgiName, fullName, prefix});
  const value = useComputed<string | number | undefined>(() => {
      if (!data.value) {
        return undefined;
      }
      const val = (mode === 'fullName' && prefix) ? data.value[prefix][fullName] : data.value[name];
      if (typeof val === "boolean") {
        return val ? "1": "0";
      }
      return val;
  });
  const prefixData = useComputed<InputFormData>(() => (mode === 'fullName' && prefix) ? (data.value[prefix] as InputFormData): data.value);
  const stringValue = useComputed<string>(() => value.value === undefined ? "" : typeof value.value === "string" ? value.value : `${value.value}`);
  const disabled = useComputed<boolean>(() => !!disabledFields?.value[name]);
  const inputClass = useComputed<string>(() => {
    let formCls = '';
    if (type === "multiselect") {
      formCls = "form-check form-switch";
    } else if (type === "checkbox") {
      formCls = "form-check-input";
    } else if(type === "select") {
      formCls = "form-select";
    } else if (title) {
      formCls = "form-control";
    }
    const err = error ? (error.value ? " is-invalid" : " is-valid") : "";
    return `${formCls}${err} ${className}`;
  });

  const inpProps: BaseInputProps = {
    name,
    type,
    title,
    disabled,
    placeholder,
    className: inputClass,
    id: `model-${name}`,
    "aria-describedby": describedBy,
    onInput: (ev: Event) => {
      const target = ev.target as HTMLInputElement;
      if (type === 'number') {
        setValue(name, parseInt(target.value, 10));
      } else if (type == 'checkbox') {
        setValue(name, target.checked);
      } else{
        setValue(name, target.value);
      }
    },
  };
  switch (type) {
    case "radio":
      return <RadioInput setValue={setValue} value={stringValue} options={options} disabledFields={disabledFields} {...inpProps} />;
    case "select":
      return <SelectInput setValue={setValue} value={stringValue} options={options} {...inpProps} />;
    case "multiselect":
      return <MultiSelectInput setValue={setValue} data={prefixData} options={options} disabledFields={disabledFields} {...inpProps} />;
    case "datalist":
      inpProps.type = datalist_type ?? 'text';
      return <DataListInput options={options} value={stringValue} {...inpProps} />;
    case 'checkbox':
      return <input checked={value.value === "1"} {...inpProps} />;
    case 'password':
      return <PasswordInput value={stringValue} allowReveal={allowReveal} {...inpProps} />;
    default:
      return <input value={value} {...inpProps} />;
  }
}
