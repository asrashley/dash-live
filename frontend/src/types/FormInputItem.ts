import { SelectOptionType } from "./SelectOptionType";

export interface FormInputItem {
    columns?: string[];
    className?: string;
    datalist_type?: 'text' | 'number';
    disabled?: boolean;
    error?: string;
    featured?: boolean;
    fullName: string;
    href?: string;
    link_title?: string;
    max?: number;
    min?: number;
    maxlength?: number;
    minlength?: number;
    multiple?: boolean;
    name: string;
    options?: SelectOptionType[];
    pattern?: string;
    placeholder?: string;
    prefix: string;
    required?: boolean;
    rowClass?: string;
    shortName: string;
    spellcheck?: boolean;
    step?: number;
    title: string;
    text?: string;
    type: 'checkbox' | 'datalist' | 'hidden' | 'multiselect' | 'number' | 'select' | 'radio' | 'link' | 'text';
    //value: boolean | number | string | null;
}