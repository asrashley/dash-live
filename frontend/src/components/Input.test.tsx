import { signal } from "@preact/signals";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../test/renderWithProviders";
import { Input, InputProps } from "./Input";
import { FormInputItem } from "../types/FormInputItem";
import { SelectOptionType } from "../types/SelectOptionType";

import allManifests from '../test/fixtures/manifests.json';
import { InputFormData } from "../types/InputFormData";

describe("Input component", () => {
  const data = signal<InputFormData>({});
  const disabledFields = signal<Record<string, boolean>>({});
  const setValue = vi.fn();

  beforeEach(() => {
    disabledFields.value = {};
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("renders a Radio button", () => {
    const playbackMode: FormInputItem = {
      name: "mode",
      shortName: "mode",
      fullName: "playbackMode",
      title: "Playback Mode",
      type: "radio",
      options: [
        {
          title: "Video On Demand (using live profile)",
          value: "vod",
        },
        {
          title: "Live stream (using live profile)",
          value: "live",
        },
        {
          title: "Video On Demand (using on-demand profile)",
          value: "odvod",
        },
      ],
    };
    const props: InputProps = {
      ...playbackMode,
      data,
      disabledFields,
      mode: 'cgi',
      setValue,
    };
    data.value = {
        mode: "vod",
    };
    disabledFields['mode__odvod'] = true;
    const { getAllBySelector, getBySelector, asFragment } = renderWithProviders(
      <Input {...props} />
    );
    const elts = getAllBySelector('input[type="radio"]');
    expect(elts.length).toEqual(playbackMode.options.length);
    playbackMode.options.forEach((opt: SelectOptionType, idx: number) => {
      const elt = elts[idx] as HTMLInputElement;
      expect(elt.value).toEqual(opt.value);
      expect(elt.name).toEqual(playbackMode.name);
      expect(elt.getAttribute("id")).toEqual(`radio-mode-${opt.value}`);
      const label = getBySelector(`label[for="radio-mode-${opt.value}"]`);
      expect(label.innerHTML).toEqual(opt.title);
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders a Select input', async () => {
    const user = userEvent.setup();
    const names = [...Object.keys(allManifests)];
    names.sort();
    const selectManifest: FormInputItem = {
        name: "manifest",
        shortName: "manifest",
        fullName: "manifest",
        title: "Manifest",
        text: "Manifest template to use",
        type: "select",
        options: names.map((name) => {
          const msft = allManifests[name];
          return {
            title: msft.title,
            value: name,
          };
        }),
      };
      const props: InputProps = {
        ...selectManifest,
        mode: 'cgi',
        data,
        disabledFields,
        setValue,
      };
      data.value = {
        manifest: names[0],
      };
      const { getAllBySelector, getBySelector, asFragment } = renderWithProviders(
        <Input {...props} />
      );
      const elts = getAllBySelector('#model-manifest option');
      expect(elts.length).toEqual(selectManifest.options.length);
      selectManifest.options.forEach((opt: SelectOptionType, idx: number) => {
        const elt = elts[idx] as HTMLInputElement;
        expect(elt.value).toEqual(opt.value);
        expect(elt.innerHTML).toEqual(opt.title);
      });
      const input = getBySelector('#model-manifest');
      await user.selectOptions(input, [names[2]]);
      expect(setValue).toHaveBeenCalledTimes(1);
      expect(setValue).toHaveBeenCalledWith('manifest', names[2]);
      expect(asFragment()).toMatchSnapshot();
    });

    test('renders a MultiSelectInput', async () => {
        const user = userEvent.setup();
        const drmSystems = ['marlin', 'playready', 'clearkey'];
        const selectDrmSystem: FormInputItem = {
            name: "drms",
            shortName: "drms",
            fullName: "drms",
            title: "DRM systems",
            text: "DRM systems to enable for encrypted AdaptationSets",
            type: "multiselect",
            options: drmSystems.map(name => ({
                name,
                title: name,
                value: name,
            })),
        };
        const props: InputProps = {
            ...selectDrmSystem,
            mode: 'cgi',
            data,
            disabledFields,
            setValue,
        };
        data.value = {
            marlin: false,
            playready: true,
            clearkey: false,
        };
        const { getAllBySelector, getBySelector, getByTestId, asFragment } = renderWithProviders(<Input {...props} />);
        getByTestId('msi-drms');
        const elts = getAllBySelector('.form-check-input');
        expect(elts.length).toEqual(drmSystems.length);
        selectDrmSystem.options.forEach((opt: SelectOptionType, idx: number) => {
            const elt = elts[idx] as HTMLInputElement;
            expect(elt.name).toEqual(opt.name);
            expect(elt.checked).toEqual(data.value[opt.name]);
            const label = getBySelector(`label[for="${elt.getAttribute('id')}"]`);
            expect(label.innerHTML).toEqual(opt.title);
        });
        await user.click(elts[0]);
        expect(setValue).toHaveBeenCalledTimes(1);
        expect(setValue).toHaveBeenLastCalledWith(drmSystems[0], true);
        await user.click(elts[2]);
        expect(setValue).toHaveBeenCalledTimes(2);
        expect(setValue).toHaveBeenLastCalledWith(drmSystems[2], true);
        expect(asFragment()).toMatchSnapshot();
    });
});
