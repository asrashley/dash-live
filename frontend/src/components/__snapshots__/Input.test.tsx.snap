// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input component > renders a MultiSelectInput 1`] = `
<DocumentFragment>
  <div
    class="form-control "
    data-testid="msi-drms"
  >
    <div
      class="form-check form-check-inline"
    >
      <input
        class="form-check-input"
        id="msimarlin"
        name="marlin"
        type="checkbox"
      />
      <label
        class="form-check-label me-3"
        for="msimarlin"
      >
        marlin
      </label>
    </div>
    <div
      class="form-check form-check-inline"
    >
      <input
        class="form-check-input"
        id="msiplayready"
        name="playready"
        type="checkbox"
      />
      <label
        class="form-check-label me-3"
        for="msiplayready"
      >
        playready
      </label>
    </div>
    <div
      class="form-check form-check-inline"
    >
      <input
        class="form-check-input"
        id="msiclearkey"
        name="clearkey"
        type="checkbox"
      />
      <label
        class="form-check-label me-3"
        for="msiclearkey"
      >
        clearkey
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input component > renders a Radio button 1`] = `
<DocumentFragment>
  <div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="radio-mode-vod"
        name="mode"
        type="radio"
        value="vod"
      />
      <label
        class="form-check-label"
        for="radio-mode-vod"
      >
        Video On Demand (using live profile)
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="radio-mode-live"
        name="mode"
        type="radio"
        value="live"
      />
      <label
        class="form-check-label"
        for="radio-mode-live"
      >
        Live stream (using live profile)
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="radio-mode-odvod"
        name="mode"
        type="radio"
        value="odvod"
      />
      <label
        class="form-check-label"
        for="radio-mode-odvod"
      >
        Video On Demand (using on-demand profile)
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input component > renders a Select input 1`] = `
<DocumentFragment>
  <select
    class="form-select "
    id="model-manifest"
    name="manifest"
    title="Manifest"
    type="select"
  >
    <option
      value="hand_made.mpd"
    >
      Hand-made manifest
    </option>
    <option
      value="manifest_a.mpd"
    >
      Vendor A live profile
    </option>
    <option
      value="manifest_b.mpd"
    >
      Vendor B VOD using live profile
    </option>
    <option
      value="manifest_e.mpd"
    >
      Vendor E live profile
    </option>
    <option
      value="manifest_ef.mpd"
    >
      Vendor EF live profile
    </option>
    <option
      value="manifest_h.mpd"
    >
      Vendor H live profile
    </option>
    <option
      value="manifest_i.mpd"
    >
      Vendor I live profile
    </option>
    <option
      value="manifest_n.mpd"
    >
      Provider N live profile
    </option>
    <option
      value="manifest_vod_aiv.mpd"
    >
      AIV on demand profile
    </option>
  </select>
</DocumentFragment>
`;
